Introduce `JUnitMethodSourceGenericParams` bug checker

mohamedsamehsalah opened the pull request:
JUnit's parameterized tests, annotated with `@MethodSource`, do not have generic type information at runtime. This means that we can pass a Set of Integers as an argument to a test method which accepts a Set of Strings, and JUnit would disregard the accepted type.

This bug checker flags such cases.

#### Potential Bug ⚠️ 
```java
  private static ImmutableSet<Arguments> demoTestCases() {
    return ImmutableSet.of(arguments(ImmutableSet.of(1)));
  }

  @ParameterizedTest
  @MethodSource("demoTestCases")
  void demo(ImmutableSet<String> setOfStrings) {
    ImmutableSet<Integer> setOfIntegers = ImmutableSet.of(1);

    assertThat(setOfStrings).isEqualTo(setOfIntegers); // true
  }
```

### Suggested commit message
```
Introduce `JUnitMethodSourceGenericParams` bug checker
```

Commit:
Introduce `JUnitMethodSourceGenericParams` bug checker

rickie reviewed (commented):
Nice check idea! Mostly testing that should be improved. I didn't go through the code yet properly, will do that after my holiday :)

Commit:
Add test coverage for #getLowerBound

Commit:
Address test mutation reports

Commit:
Add a test for MoreJUniMatchers#findMethodSourceAnnotation

Commit:
Introduce `IrrelevantNonParameterizedType`

mohamedsamehsalah reviewed (commented):
Thanks for review @rickie 🙌 

I tried to address the mutation test reports, but I will need help to cover all of them. I tried to put the reasoning in the comments.

rickie reviewed (commented):
Only some small comments so far. Takes a while to review, but flashing already.

rickie committed:
Suggestions

rickie reviewed (commented):
Lol I confused myself with flushing what I had and not 😬.
Pushing a commit with some changes. Let's discuss some of the mutants to get that more sorted out :). 

The check has a nice shape. Let's do these two things:
- Run it on a repo of ours internally (I recall you might have done this a while back as you asked me about it 🤔).
- I've to continue reviewing :)

rickie committed:
Fix build

rickie committed:
Attempt to kill some mutatnts

rickie committed:
Simplify MoreASTHelpers method a bit

rickie committed:
Attempt to kill more mutants

rickie committed:
Revert "Attempt to kill more mutants"

This reverts commit 44ed06333ab5609cfbad9f860e2c5a1fd86908fa.

rickie committed:
Address XXX review comments on JUnitMethodSourceGenericParams

Major refactorings:
- Removed Optional wrapper from offending providers list
- Replaced index-loop with Streams.zip in hasApplicableTypes
- Renamed methods: getPotentialOffendingProvider → collectPotentialOffendingProviders,
  findMethodByName → hasMethodName, getTypes → toTypes, findMembers → extractMembers
- Reordered VisitorState parameter to last position in addPotentialOffendingProvider
- Switched getReturnExpressions to use ImmutableList.Builder
- Converted nested if-else to switch expressions
- Rewrote Javadoc and improved code clarity throughout
- Added test coverage for captured TYPEVAR branch in MoreASTHelpersTest

rickie committed:
Fix multi-type-arg generic compatibility check and tighten internals

**Correctness fix:** The `allParameterisedArgumentsHaveApplicableTypes` method
previously returned `true` as soon as ONE type argument matched instead of
requiring ALL type arguments to match. This allowed a `Map<String, Boolean>` to
silently satisfy a `Map<String, Integer>` parameter via the matching key type.
Fixed by extracting an `isApplicableTypeArgument` predicate and using
`Streams.zip(...).allMatch(...)` to ensure all type arguments are checked.

**Test coverage:** Added two multi-type-argument test cases (one asserting
flagging in `mapValueMismatchIdentificationTestCases`, one for the no-false-
positive path in `mapNoIdentificationTestCases`). These would fail against the
previous buggy semantics.

**Supporting refactors:**
- Renamed `findOffendingMethodSourceProvider` (returned `Optional<MethodTree>`
  with no effect) to `isOffendingMethodSourceProvider` (pure boolean predicate).
  Caller simplified from map/filter pattern to just `.filter(...)`.
- Replaced recursive `buildPotentialOffendingProviders` + mutable
  `collectPotentialOffendingProviders` + `addPotentialOffendingProvider` with
  pure recursive `findPotentialOffendingProviders` returning `ImmutableList`.
- Simplified `addPotentialOffendingProvider`'s conditional skip logic with a
  computed `int skip` variable.
- Applied systematic naming sweep: `getOffendingMethodSourceProviders` →
  `findOffendingMethodSourceProviders`; `getReturnExpressions` →
  `collectReturnExpressions`; `getMethodNames` → `methodNames`;
  `ARGUMENT_SET` → `ARGUMENT_SET_FACTORY`; method parameter `methodName` →
  `factoryReference` (with clarifying Javadoc on the `lastIndexOf('#')` trick).

The `extractMembers` Stream-return deferred per prior agreement.

rickie committed:
Tiny improvement

rickie committed:
Merge branch 'master' into mohamedsamehsalah/junit-method-source-generic-params

rickie committed:
Simplify internals and broaden test coverage

Drop the single-field `PotentialOffendingArgumentsProvider` record and
replace the recursive `extractMembers` stream helper with a `TreeScanner`
that already descends into nested classes. Fix the misspelled
`allParameterised...` method, the dangling private-field reference in
the class Javadoc, and an "iff" that wasn't biconditional. Add XXX
comments documenting two mutants that are provably/structurally
unkillable.

Tests: rename `runtimeValueTestCases` to make clear the checker does
flag it, hook up the previously-defined `noIdentificationCollection`
factory, reorder negatives-first, and add four targeted cases that
kill surviving mutants in the size guard, the non-`Arguments` filter,
the non-wildcard supertype branch, and the non-parameterized early
return.

rickie committed:
Merge branch 'mohamedsamehsalah/junit-method-source-generic-params' of github.com:PicnicSupermarket/error-prone-support into mohamedsamehsalah/junit-method-source-generic-params
